lib/statefileio.c

```diff
@@ -106,11 +106,41 @@ sf_FIST_delay_on_write();
 //
 
 
+enum {
+    DEVICE_TYPE_NORMAL = 0,
+    DEVICE_TYPE_NBD = 1
+};
+
+static int
+get_device_type(
+      const char *path)
+{
+    static const int nbd_major = 43;
+
+    struct stat stats;
+    if (stat(path, &stats) == -1)
+    {
+        log_internal(MTC_LOG_ERR, "SF: failed to execute stat (sys %d).\n", errno);
+        return DEVICE_TYPE_NORMAL;
+    }
+    if (S_ISBLK(stats.st_mode) && major(stats.st_rdev) == nbd_major) {
+        return DEVICE_TYPE_NBD;
+    }
+    return DEVICE_TYPE_NORMAL;
+}
+
 extern int
 sf_open(
     char *path)
 {
-    return open(path, (O_RDWR | O_DIRECT));
+    const int device_type = get_device_type(path);
+    const int flags = O_RDWR | O_DIRECT | (device_type == DEVICE_TYPE_NBD ? O_SYNC : 0);
+
+    int fd;
+    do {
+        fd = open(path, flags);
+    } while (fd < 0 && (errno == EAGAIN || errno == EINTR));
```

Comment:
[posix docs for open
](https://pubs.opengroup.org/onlinepubs/007904875/functions/open.html) say for EAGAIN: "The path argument names the slave side of a pseudo-terminal device that is locked".
And the Linux one doesn't mention EAGAIN at all (to get EAGAIN you'd usually have to get the file descriptor into non-blocking mode, but you are opening the file here so you didn't have a chance yet).

Did you observe 'open' returning EAGAIN on a system? I'd prefer to treat it as an error and propagate the error, because we do not know why EAGAIN was returned, and whether retrying would work or whether it'd result in an infinite loop.